no-undefined-references: add support for regex as option

sisp opened the pull request:
<!--
  Please check the needed checkboxes ([ ] -> [x]). Leave the
  comments as they are, they won’t show on GitHub.
  We are excited about pull requests, but please try to limit the scope, provide
  a general description of the changes, and remember, it’s up to you to convince
  us to land it.
-->

### Initial checklist

*   [x] I read the support docs <!-- https://github.com/remarkjs/.github/blob/main/support.md -->
*   [x] I read the contributing guide <!-- https://github.com/remarkjs/.github/blob/main/contributing.md -->
*   [x] I agree to follow the code of conduct <!-- https://github.com/remarkjs/.github/blob/main/code-of-conduct.md -->
*   [x] I searched issues and couldn’t find anything (or linked relevant results below) <!-- https://github.com/search?q=user%3Aremarkjs&type=Issues -->
*   [x] If applicable, I’ve added docs and tests

### Description of changes

I've added support for regex patterns in the `allow` option in the [`no-undefined-references`](https://github.com/remarkjs/remark-lint/tree/main/packages/remark-lint-no-undefined-references) rule.

Let me know what you think. :slightly_smiling_face:

Closes #296.

<!--do not edit: pr-->

sisp commented:
Hm, why is this bad?

> Error: /home/runner/work/remark-lint/remark-lint/packages/remark-lint-no-undefined-references/index.js: line 347, col 55, Error - Do not pass function `isString` directly to `.filter(…)`. (unicorn/no-array-callback-reference)
> Error: /home/runner/work/remark-lint/remark-lint/packages/remark-lint-no-undefined-references/index.js: line 352, col 47, Error - Do not pass function `isRegExp` directly to `.filter(…)`. (unicorn/no-array-callback-reference)

Passing this custom type guard to `.filter(...)` directly types the return value properly. :thinking:

wooorm commented on packages/remark-lint-no-undefined-references/index.js:
```diff
@@ -8,9 +8,10 @@
+ *     *   `allow` (`Array<string | RegExp | { source: string }>`,
+ *         default: `[]`)
+ *         — text or a regex that you want to be allowed between `[` and `]`
+ *         even though it’s undefined
```
- Needs to explain the `source` object version
- Probably also needs to explain how regexes are handled (case-insensitive?)

sisp replied:
See db695ae40f39dd1014656cc9eb6792c473077683. WDYT?

wooorm commented on packages/remark-lint-no-undefined-references/index.js:
```diff
@@ -61,6 +62,17 @@
+ *   [baz][b]
+ *
+ *   Matching is case-insensitive: [bar][PKG.SUBPKG]
+ *
```
I think it’s better to have one example that fails here (`b`?)

And maybe instead of `pkg.subpkg` here, a value of `c.d` might make more sense? Conveys the same, while still using abstract values

sisp replied:
> I think it’s better to have one example that fails here (`b`?)

See 73a7b0e1d2829540843310848ee66080d3266d0b. WDYT?

> And maybe instead of `pkg.subpkg` here, a value of `c.d` might make more sense? Conveys the same, while still using abstract values

See ae0f12b00ceb516fbb649c8ab00741891a600d1a. WDYT?

wooorm commented on packages/remark-lint-no-undefined-references/index.js:
```diff
@@ -123,9 +137,7 @@ const remarkLintNoUndefinedReferences = lintRule(
-    const allow = new Set(
-      (option.allow || []).map((d) => normalizeIdentifier(d))
-    )
+    const allow = new AllowSet(option.allow || [])
```
I don’t think a whole custom class is needed here. We very rarely use classes.

wooorm replied:
E.g., how about this:

```suggestion
/** @type {Array<RegExp>} */
const regexes = [];
/** @type {Array<string>} */
const strings = []

let index = -1;

while (++index < allow.length) {
    const value = allow[index];
    if (typeof value === 'string') {
        strings.push(value)
    } else {
        regexes.push(new RegExp(value.source, 'i'))
    }
}
```

sisp replied:
The class creates an abstraction around checking whether an identifier matches one of the allow values which isn't simple a simple set membership check anymore. The check occurs in two places, so I thought it would be nice to hide the checking logic.

Would you rather prefer your suggestion plus a closure below your suggested snippet that abstracts the check?

wooorm replied:
Yes, I understand the need to abstract the check itself. Indeed, I’d prefer a closure below!

sisp replied:
See 8c27df2dc6820b9a6bb67eba7f3dc1a49f9ab6da. WDYT?

wooorm commented:
> Hm, why is this bad?

Because it leads to bugs. `map`, `filter`, etc, pass 2nd and 3rd parameters to functions.
Sometimes, functions accept 2nd or 3rd parameters, or variadic parameters.
This can be hard to remember and leads to bugs, particularly a viral twitter thread every other month 😅

More info: https://github.com/sindresorhus/eslint-plugin-unicorn/blob/main/docs/rules/no-array-callback-reference.md

sisp commented:
I believe I've addressed your review feedback, @wooorm. Would you mind checking again? :pray:

sisp commented:
Now I'm not sure where this error is coming from:

> ERROR: Coverage for lines (99.98%) does not meet global threshold (100%)
ERROR: Coverage for branches (99.88%) does not meet global threshold (100%)
ERROR: Coverage for statements (99.98%) does not meet global threshold (100%)
Error: Process completed with exit code 1.

wooorm commented:
Well, looks like there’s an `any`!

Did you try and `cd` into the package you are working on, and run tests there? In this case, it should be the build step. So, `npm run build` checks just that one package’s build step.

sisp commented:
Thanks for the hint! Yes, one line isn't covered which is related to the JS-native `RegExp` case:

https://github.com/remarkjs/remark-lint/blob/5e5bd15ff003b967e9fd88c8864d6cb36e0db9c1/packages/remark-lint-no-undefined-references/index.js#L167

Is there any way to implement a test for this case? The example-based tests don't seem to support JS.

wooorm commented on packages/remark-lint-no-undefined-references/index.js:
```diff
@@ -305,12 +346,26 @@ const remarkLintNoUndefinedReferences = lintRule(
+     * @returns {boolean}
+     */
+    function isAllowed(id) {
+      if (strings.size > 0 && strings.has(normalizeIdentifier(id))) {
+        return true
+      }
+      if (regexes.some((regex) => regex.test(id))) {
+        return true
+      }
+      return false
```
Everything can be replaced with:

```suggestion
      const normalized = normalizeIdentifier(id)
      return strings.has(normalized) || regexes.some((regex) => regex.test(normalized))
```

*   I think `normalized` should be used for the regex too.
*   And I don’t think the `set.size` check is needed?

sisp replied:
> * I think normalized should be used for the regex too.

Hm, I had thought about it but then concluded that coming up with the right regex would require internal knowledge about the `normalizeIdentifier(...)` function whereas for the string case it doesn't because both the string and the ID are normalized. WDYT?

> * And I don’t think the `set.size` check is needed?

I added this check to avoid an unnecessay call to `normalizeIdentifier(...)` in case the `strings` set is empty. If we conclude that the regex tests should receive normalized IDs, then you're right. If the regex tests should receive _unnormalized_ IDs, the above could still be "simplified" to

```js
return (
  (strings.size > 0 && strings.has(normalizeIdentifier(id))) ||
  regexes.some((regex) => regex.test(id))
)
```

although I find the formatting produced by Prettier less readable TBH. But it's your call in the end.

wooorm replied:
responded below in a whole review

wooorm reviewed (commented):
> The example-based tests don't seem to support JS.

True, it doesn’t.

You could maybe add a small test around here?
https://github.com/remarkjs/remark-lint/blob/37041219cb6e72c321a6179e77c7083c3216bcfa/test.js#L221

sisp commented:
> > The example-based tests don't seem to support JS.
> 
> True, it doesn’t.
> 
> You could maybe add a small test around here?
> 
> https://github.com/remarkjs/remark-lint/blob/37041219cb6e72c321a6179e77c7083c3216bcfa/test.js#L221

See 572f0d8a701da156a6f3fe622a4f59befbaaa4b4. WDYT?

wooorm reviewed (commented):
This looks good! I am still in favor of similifying the check by using `normalize` for both strings and regexes. Normalize is very simple, basically case-insensitive and normalizing white space, I am assuming users of regexes will match it intuitively  https://github.com/micromark/micromark/blob/main/packages/micromark-util-normalize-identifier/dev/index.js

sisp commented:
Alright, I've simplified `isAllowed` as you suggested. I also noticed that there were some minor Prettier formatting errors that I've fixed, too.

Unless you have other requests, I think this PR is ready to be merged. :tada:

wooorm commented:
Thanks @sisp! [Released](https://github.com/remarkjs/remark-lint/releases/tag/remark-lint-no-undefined-references%404.2.0)!

sisp commented:
Awesome! 🎉 Thanks for the thorough review and feedback, @wooorm! 🙏
